providers/oauth/oauth.httpobserver.js

```diff
@@ -75,12 +84,26 @@ FirefoxTabsAuth.prototype.launchAuthFlow = function (authUrl, stateObj, continua
     }, 100);
   };
 
+  fail = function () {
+    gBrowser.removeProgressListener(listener);
+    gBrowser.removeTab(tab);
+    continuation(undefined, 'Error in launchAuthFlow');
+  };
+
   httpRequestObserver.register();
 
   //Timeout before switching makes sure the http observer is installed.
   setTimeout(function () {
     tab = gBrowser.addTab(authUrl);
-    gBrowser.selectedTab = tab;
+    if (interactive) {
+      gBrowser.selectedTab = tab;
+    } else {
+      setTimeout(function () {
+        if (!hasCredentials) {
+          fail();
+        }
+      }, 5000);
```

Comment:
Just want to check the reasoning for why 5000ms is the appropriate timeout here - is this something that could potentially block the user for this long?